src/Chainweb/BlockHeader.hs

```diff
@@ -211,19 +233,172 @@ encodeEpochStartTime (EpochStartTime t) = encodeTime t
+            -- of the block. The value must be strictly monotonically increasing
+            -- within the chain of blocks. Nodes must ignore blocks with values
+            -- that are in the future and reconsider a block when its value is
+            -- in the past. Nodes do not have to store blocks until they become
+            -- recent (but may do it).
+            --
+            -- The block creation time is used to determine the block difficulty for
+            -- future blocks.
+            --
+            -- Nodes are not supposed to consider the creation time when
+            -- choosing between two valid (this implies that creation time of a
+            -- block is not the future) forks.
+            --
+            -- This creates an incentive for nodes to maintain an accurate clock
+            -- with respect to an (unspecified) commonly accepted time source,
+            -- such as the public NTP network.
+            --
+            -- It is possible that a miner always chooses the smallest possible
+            -- creation time value. It is not clear what advantage a miner would
+            -- gain from doing so, but attack models should consider and
+            -- investigate such behavior.
+            --
+            -- On the other hand miners may choose to compute forks with creation
+            -- time long in the future. By doing so, the difficulty on such a fork
+            -- would decrease allowing the miner to compute very long chains very
+            -- quickly. However, those chains would become valid only after a long
+            -- time passed and would be of low PoW weight. The algorithm for
+            -- computing the difficulty must ensure this strategy doesn't give
+            -- an advantage to an attacker that would increase the success
+            -- probability for an attack.
+
+        , _blockParent :: {-# UNPACK #-} !BlockHash
+            -- ^ authoritative
+
+        , _blockAdjacentHashes :: !BlockHashRecord
+            -- ^ authoritative
+
+        , _blockTarget :: {-# UNPACK #-} !HashTarget
+            -- ^ authoritative
+
+        , _blockPayloadHash :: {-# UNPACK #-} !BlockPayloadHash
+            -- ^ authoritative
+
+        , _blockChainId :: {-# UNPACK #-} !ChainId
+
+        , _blockWeight :: {-# UNPACK #-} !BlockWeight
+            -- ^ the accumulated weight of the chain. It is redundant information
+            -- that is subject to the inductive property that the block weight
+            -- of a block is the block weight of the parent plus the difficulty
+            -- of the block.
+
+        , _blockHeight :: {-# UNPACK #-} !BlockHeight
+            -- ^ block height records the length of the chain. It is redundant
+            -- information and thus subject the inductive property that
+            -- the block height of a block is the block height of its parent
+            -- plus one.
+
+        , _blockChainwebVersion :: !ChainwebVersion
```

Comment:
Consider `ChainwebVersionCode` and define an `HasChainwebVersion` instance for `ChainwebVersionCode`.